lib/RWMutex.ts

```diff
@@ -0,0 +1,162 @@
+function timeoutPromise(delay) {
+  return new Promise((resolve) => {
+    setTimeout(() => {
+      resolve();
+    }, delay);
+  });
+}
+
+/*
+ * Client representation of a mutex in the database
+ */
+export default class RWMutex {
+  _coll;
+  _lockID;
+  _clientID;
+  _options;
+
+  constructor(coll, lockID, clientID, options = { sleepTime: 5000 }) {
+    this._coll = coll;
+    // TODO: typecheck these
+    this._lockID = lockID;
+    this._clientID = clientID;
+    this._options = options;
+  }
+
+  async lock() {
+    // In a loop, do the following:
+    // 1. Check clientID of the lock. If lock.clientID > this.clientID, bail
+    // 2. Acquire the lock
+    while (true) {
+      // Find the lock
+      let lock;
+      try {
+        // find the lock
+        lock = await this._coll.find({ lockID: this._lockID }).limit(1).next();
+        if (!lock) {
+          // lock doesn't exist yet, so we should create it. Empty string is considered
+          // lexographically < any other string, so we can use that as the zero value
+          await this._coll.insert({
+            lockID: this._lockID,
+            readers: [],
+            writer: "",
+          });
+          continue;
+        }
+      } catch (err) {
+        throw new Error(`error finding lock: ${err.message}`);
+      }
+
+      // if this clientID already has the lock, we re-enter the lock and return
+      if (lock.writer === this._clientID) {
+        return;
+      }
+
+      // Acquire the lock
+      try {
+        const result = await this._coll.updateOne({
+          lockID: this._lockID,
+          readers: [],
+          writer: "",
+        }, {
+          $set: {
+            writer: this._clientID,
+          },
+        });
+        if (result.matchedCount > 0) {
+          return true;
+        }
+      } catch (err) {
+        throw new Error(`error aquiring lock: ${err.message}`);
+      }
+
+      await timeoutPromise(this._options.sleepTime);
+    }
+  }
+
+  async unlock() {
+    let result;
+    try {
+      result = await this._coll.updateOne({
+        lockID: this._lockID,
+        writer: this._clientID,
+      }, {
+        $set: {
```

Comment:
we should note that we don't remove locks when they're no longer used. This means the collection will grow forever. Not as issue for our use case, but one we should document since this is public